oeps/oep-0031-arch-i18n.rst

```diff
@@ -0,0 +1,122 @@
+==============================
+OEP-0031: Internationalization
+==============================
+
+.. list-table::
+
+   * - OEP
+     - :doc:`OEP-0031 <oep-0031-arch-i18n>`
+   * - Title
+     - Internationalization
+   * - Last Modified
+     - 2019-01-07
+   * - Authors
+     -
+   * - Arbiter
+     -
+   * - Status
+     - Draft
+   * - Type
+     - Architecture
+   * - Created
+     - 2019-01-07
+   * - Review Period
+     -
+
+.. contents::
+   :local:
+   :depth: 2
+
+Motivation
+==========
+
+Internationalization (i18n) is a vital part of unlocking content for users around the globe and effectively supporting our business partners. In order to help us reuse localized components from app to app, the different parts of the edX codebase should align on a consistent handling of i18n. This document covers our approach to internationalizing new front-end applications and components written in React.
+
+In this OEP we use the terms locale and language more or less interchangeably. However, note that this document addresses primarily making code multilingual, i.e., the issue of presenting the same content in different languages. Considering ways to make it multi-regional, to allow for presenting different content for different countries, is out of scope for this OEP.
+
+Technology Choice
+=================
+
+We will use `react-intl <https://github.com/yahoo/react-intl>`_ to internationalize React components, in agreement with `Paragon ADR-2 <https://github.com/edx/paragon/blob/f5b9f876800e5154ef01d04f682b8cda21b3e439/docs/decisions/0002-react-i18n.rst>`_. react-intl is in line with current industry standards in i18n and satisfies all currently identified i18n needs, including multiple plurals.
+
+Alternatives Considered
+-----------------------
+
+An alternative choice we considered was `i18next <https://react.i18next.com/>`_. **[TODO]**
+
+Locale
+======
+
+Localized content for new micro frontends should be at a URL that includes the language as the first part of the path after the domain. (Example: ``account.edx.org/es/``) The language string should be in the form of a `BCP 47 language tag <http://tools.ietf.org/html/rfc5646>`_. If the user enters the site at a URL that doesn't include a language string, we will determine a locale and then forward them to the corresponding URL. For consistency, this applies even if their language is English: ``account.edx.org/`` should be forwarded to ``account.edx.org/en/`` . We will use the most general, two-letter language codes to start with, and re-evaluate as necessary.
```

Comment:
Chiming in without full context here, so forgive me if this is naive.  Does the two-letter language code inform number formatting in localizations, for instance?  Might be worth going straight to en_US, en_UK, en_CA etc... it feels like it's likely to become a stumbling block almost immediately.  If not determined by the URL path, by what criteria do we determine which country-specific localizations to use?

Reply:
If our preference is two character lang we should make sure we have a common function to convert en-US, en-GB to en.  I think when we are talking China we are going to have to make decisions re: 5 character codes zh for traditional and simplified as supporting both maybe an ask

Reply:
It looks like that takes us out of scope for the oep: https://github.com/edx/open-edx-proposals/pull/100/files#diff-00db08b214255fa07b73c64617cebe2dR35 (although I'm worried we may want to rename the oep to specify that it's not about Internationalization writ large, given that omission).

Reply:
These are all really good points.  I've punted discussion of locale codes to a future update to this OEP.